rohmu/object_storage/s3.py

```diff
@@ -24,16 +24,50 @@
 )
 from botocore.response import StreamingBody
 from rohmu.util import batched
-from typing import Any, BinaryIO, Collection, Iterator, Optional, Tuple, Union
+from typing import Any, BinaryIO, cast, Collection, Iterator, Optional, Tuple, TYPE_CHECKING, Union
 
-import boto3
 import botocore.client
 import botocore.config
 import botocore.exceptions
 import botocore.session
 import math
 import time
 
+# botocore typing stubs are incomplete. We either have to write all the stubs we need
+# or work around the problem a bit by importing the typing library only for type checking purposes.
+if TYPE_CHECKING:
+    from mypy_boto3_s3 import S3Client
+
+    s3_client_type = S3Client
+else:
+    s3_client_type = Any
+
+
+def create_s3_client(
+    *,
+    session: botocore.session.Session,
+    config: botocore.config.Config,
+    aws_access_key_id: Optional[str],
+    aws_secret_access_key: Optional[str],
+    aws_session_token: Optional[str],
+    region_name: str,
+    verify: Optional[bool | str] = None,
+    endpoint_url: Optional[str] = None,
+) -> s3_client_type:
+    return cast(
```

Comment:
cast is not relevant for non-type-checking case anyway; this whole block could happen under if TYPE_CHECKING? And return value could be always S3Client (even if it is not defined).

Reply:
Because this needs `mypy_boto3_s3` there will have to be a conditional import and cast. Alternatively we will have to install `mypy_boto3_s3` during runtime as well and that seems wrong for this scenario. I did tidy up the logic a bit more so that no cast is required at runtime and no special type variables are defined